lib/spack/spack/test/cmd_extensions.py

```diff
@@ -109,3 +154,16 @@ def test_command_with_import(hello_world_with_module_in_root):
     assert 'Hello folks!' in output
     output = hello_world_with_module_in_root('global')
     assert 'bar' in output
+
+
+def test_sub_missing_cmd(hello_world_sub_no_cmd):
```

Comment:
These tests verify correct normal and exceptional behavior of `get_[command_]module_from()` which is the function provided to make it possible for command extensions to load sub-commands implemented as modules. The "used in the sub-command itself" seems irrelevant to me: you're unit-testing the function as it is intended to be used—the macro behavior it enables is not required to be tested here.

I have not included the full, "this is how you would use all this in a command extension," because I'm not sure yet if there is one, "ideal way." The implementation in SpackDev's top-level command looks like this:
```python
SPACKDEV_FILE = os.path.realpath(os.path.expanduser(__file__))
SPACKDEV_LIB = os.path.realpath(os.path.join(os.path.dirname(SPACKDEV_FILE),
                                             '..', '..', 'lib'))
sys.path.insert(0, SPACKDEV_LIB)

description = 'develop multiple Spack packages simultaneously'
section = 'developer'
level = 'long'

_subcmd_dir = os.path.join(SPACKDEV_LIB, 'fnal', 'spack', 'dev', 'cmd')
_subcmds = None
_subcmd_functions = {}


def add_subcommand(subparser, subcmd):
    pname = spack.cmd.python_name(subcmd)
    module = spack.cmd.get_command_module_from(pname, 'fnal.spack.dev')
    sp = subparser.add_parser(subcmd, help=module.description)
    module.setup_parser(sp)
    global _subcmd_functions
    _subcmd_functions[subcmd] = getattr(module, pname)


def setup_parser(subparser):
    sp = subparser.add_subparsers(metavar='SUBCOMMAND', dest='dev_command')
    global _subcmds
    _subcmds = spack.cmd.find_commands(_subcmd_dir)
    for subcmd in _subcmds:
        add_subcommand(sp, subcmd)


def dev(parser, args):
    _subcmd_functions[args.dev_command](parser, args)
```
As you can see, it's a little disjoint. I like to package it as a separate PR eventually, but I think we need to collect a few more uses in order to understand how to package it simply.